File: cpp/bench_alloc.cpp
#include <string>
#include <iostream>
#include <chrono>

// Return by value triggers allocation (unless SSO works, so we make it large enough to bypass SSO if we want true malloc, but let's test small string creation first as user asked "making objects")
// Actually, Python allocates for EVERYTHING. C++ SSO optimization is a valid advantage.
// Let's make it comparable to Python: "prefix_" + number.
// This is likely small enough for SSO in some implementations (16 chars), but let's see.

std::string make_obj(int i) {
    return "prefix_" + std::to_string(i);
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    
    for (int i = 0; i < 1000000; ++i) {
        std::string t = make_obj(i);
        asm volatile("" :: "r"(&t) : "memory");
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;
    std::cout << "Time: " << diff.count() << " s" << std::endl;
    return 0;
}

File: rust/bench_alloc.rs
use std::time::Instant;
use std::hint::black_box;

fn make_obj(i: i32) -> String {
    format!("prefix_{}", i)
}

fn main() {
    let start = Instant::now();
    for i in 0..1_000_000 {
        let t = make_obj(i);
        black_box(t);
    }
    let duration = start.elapsed();
    println!("Time: {:.6} s", duration.as_secs_f64());
}
